proposed/event-dispatcher.md

```diff
@@ -66,54 +54,28 @@ A Listener MUST have a `void` return, and SHOULD type hint that return explicitl
 
-Listeners for a Notify event MUST NOT assume any awareness of other Listeners that may or may not be called.  A Notify Dispatcher MAY take additional steps to enforce such separation but is not required to.  (For example, cloning the message before passing it to each listener.)
+## Dispatcher
 
-A Notifier
-
-* MAY call relevant Listeners in any order it wishes.
-* MAY delay calling Listeners until some later point, such as using a queue system.
-* MAY call multiple Listeners concurrently (such as in an asynchronous system or a queue system with multiple worker processes).
-
-An Emitter calling a Notifier MUST NOT assume that any listeners have fired yet by the time the Notifier has returned.
-
-A Notifier MUST allow for the case where zero Listeners are found and MUST NOT generate an error condition in that case.
-
-### Error handling
+A Dispatcher is a service object implementing `EventDispatcherInterface`.  It is responsible for invoking listeners provided by a Listener Provider on an Event.
 
-A Notifier MUST ensure that any `\Throwable` generated by a listener does not impact or prevent other listeners from triggering normally.  It is RECOMMENDED that the Notifier log the error event but the only requirement is that one listener's exception MUST NOT block another listener.
-
-## Processor
-
-A Processor is a service object implementing `TaskProcessorInterface`.  It is used for cases where an Emitter wishes to provide data to listeners and receive data back from them.  Examples of Task use cases include:
-
-* Passing an object to a series of Listeners to allow it to be modified before it is saved to a persistence system.
-* Passing a collection to a series of Listeners to allow them to register values with it so that the Emitter may act on all of the collected information.
-* Passing a collection to a series of Listeners to allow them to modify the collection in some way before the Emitter takes action.
-* Passing some contextual information to a series of Listeners so that all of them may "vote" on what action to take, with the Emitter deciding based on the aggregate information provided.
-* Passing an object to a series of listeners and allowing one of them to set a value and then prevent further listeners from running.
-
-Tasks passed to a Processor SHOULD have some sort of mutator methods on them to allow Listeners to modify the object.  The nature of those methods is up to each implementation to determine.
-
-A Processor
+A Dispatcher
 
 * MUST call Listeners synchronously in the order they are returned from a ListenerProvider.
-* MUST return the Task it was passed after it is done invoking Listeners.
+* MUST return the same Event object it was passed after it is done invoking Listeners.
```

Comment:
Question: should this be the same event instance passed to the dispatcher, or the same instance the dispatcher passed to listeners? The latter would free up a scenario such as seen in symfony/messenger whereby the messenger may wrap the message in an envelope prior to dispatching it. If it then returned the envelope, it would be in violation of the spec per the above, but if we changed the wording as I suggest, it would still be compliant.

Reply:
That is something we should discuss IMHO, as it is unclear in my skunk-work as well, what is the proper return value.

Reply:
I would favor same object, period.  As in:

```php
$e = new Event();
$e2 = $d->dispatch($e);
assert($e === $e2);
```

Returning a wrapping envelope object means I have to either known how to unwrap it, or I need to know what its signature is.  And if that's not consistent between all implementations I now have an implementation dependency.

IMO the Symfony Message component is still "out of scope" as just not a thing we're dealing with, period.

Reply:
I tend to agree with @Crell here - he's saying essentially the same thing I mentioned when we discussed envelopes before, which is that if they are needed, they should decorate the message _before_ you pass it to the dispatcher. That way you don't have to test what was returned; you already know what it is. It also means if you call methods on that object after dispatching the event, you know you're calling methods on the same object sent to listeners.

As such, I'll retract my question and suggestion. :smile: